### PR TITLE
test moving functionality from component

### PR DESCRIPTION
**Pull request by RavisankarSelvaraju**
Create a class that contains the transformation calculation currently implemented in the control-orogen-robot_frames

### CODE REVIEW

File: src/RobotFrames.hpp [hunk 1]
```diff
@@ -210,6 +211,48 @@ class TransformationCalculator
...
+    
+    void clear_and_resize_vectors();
+    void unpack_joints(const base::samples::Joints& joint_state,
+                        const std::vector<std::string>& involved_joints,
+                        KDL::JntArray& joint_array);
+
+    /* Mimics the functionality of ConfigureHook of control-orogon-robot_frames 
+       Configure the parameters needed to calculate the transformations in the given chain
+    */
+    void configure();
```
Comment (maltewi) [inline src/RobotFrames.hpp]: when moving this part into library there's no need to stick to rock's component life-cylce. In fact I think it puts extra complexity that not required for a lib-api.

The `initialize` and `configure` functions can be merged together and put into the constructor for `ChainTransformationCalculator`.

File: src/RobotFrames.hpp [hunk 2]
```diff
@@ -210,6 +211,48 @@ class TransformationCalculator
...
+    std::vector<std::vector<std::string> >involved_active_joints_;
+    KDL::Tree tree_;
+    int n_defined_chains_;
+
+public:
+
+    void initilize(std::vector<robot_frames::Chain> chain, std::string urdf, base::samples::Joints jointstate);
+    
+    void clear_and_resize_vectors();
+    void unpack_joints(const base::samples::Joints& joint_state,
```
Comment (maltewi) [inline src/RobotFrames.hpp]: No need for being exposed as public, or? I'd prefer to make it protected then. I guess with the proposed signature this function does not modify and member variables?!?
If so, consider moving it out of the class or make it static.

**Review by maltewi — COMMENTED** [review]
I have no objections with your idea of moving this part from the orogen component to the lib.
Please have a look at the additional comments for improving the library API.

File: src/RobotFrames.hpp (hunk 2 above)
Comment (RavisankarSelvaraju) [reply src/RobotFrames.hpp]: unpack_joints() function wont modify member variables but clear_and_resize_vectors() will. 
yes, it can be protected.

File: src/RobotFrames.hpp (hunk 1 above)
Comment (RavisankarSelvaraju) [reply src/RobotFrames.hpp]: okay

File: src/RobotFrames.cpp [hunk 3]
```diff
@@ -267,4 +266,181 @@ bool TransformationCalculator::get_transform_by_joint_name(const std::string& j_
...
+        st = tree_.getChain(root, tip, chains_[i]);
+        if(!st){
+            LOG_ERROR("Error extracting chain '%s' with root '%s' and tip '%s'. The urdf_file is: %s",
+                      name.c_str(), root.c_str(), tip.c_str(), urdf_file_path.c_str());
+            throw std::runtime_error("Error extracting chain");
+        }
+        chain_names_[i] = name;
+    }
+
+    //Determine invloved joits, prepare frames storage and solvers
```
Comment (maltewi) [inline src/RobotFrames.cpp]: typo

File: src/RobotFrames.hpp [hunk 4]
```diff
@@ -8,7 +8,11 @@
 #include "base/samples/Joints.hpp"
 #include "base/samples/RigidBodyState.hpp"
 #include "kdl/tree.hpp"
-#include "robot_framesTypes.hpp"
+#include "robot_frames/robot_framesTypes.hpp"
+
+#include <kdl/chainfksolvervel_recursive.hpp>
```
Comment (maltewi) [inline src/RobotFrames.hpp]: not sure. Is  `chainfksolvervel_recursive.hpp` really needed?

**Review by maltewi — COMMENTED** [review]
did not run any tests. did you?

**Comment by RavisankarSelvaraju** [comment]
> did not run any tests. did you?

I ran a test using the component and by passing the chain in config file(.yml) i can get the transformation of the chain (i could see the transformation in rock-display). i even moved the robot in simualtion using virtual joystick and the transformation changes

File: src/RobotFrames.hpp (hunk 4 above)
Comment (RavisankarSelvaraju) [reply src/RobotFrames.hpp]: removed

File: src/RobotFrames.cpp (hunk 3 above)
Comment (RavisankarSelvaraju) [reply src/RobotFrames.cpp]: fixed

**Comment by maltewi** [comment]
Looks good to me. Could you please add a test case to the tests-subfolder testing the new `ChainTransformationCalculator`.

You can look at https://github.com/rock-control/control-robot_frames/blob/master/test/test_RobotFrames.cpp for an example. No need to be overly exhaustive there but a call to `update_transforms` with 0 chains, 1 chain and 2 chains to check the corner case would be nice.

The cmake-flag `ROCK_TEST_ENABLED` need to be enabled for compiling the unit tests (e.g. via `cmake -D ROCK_TEST_ENABLED=1 ..` in the build directory)

Please run a reconfiguration test manually once (see comment in component's merge request) by applying another config to the running component (e.g. by modifying your config in the cnd inbetween two rock-launch-calls)

**Comment by RavisankarSelvaraju** [comment]
manual reconfiguration test status:
i tried removing the dynamics port
1. i launch a CND with config 1 and got the ports generated 
2. when i launch another CND with config 2, i can get the port but also getting an error 

```
[20240209-13:53:26:220] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.state on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
[20240209-13:53:34:440] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.state on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
[20240209-13:53:34:440] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.coyote3_col_wheel_front_left_01 on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
[20240209-13:53:34:440] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.coyote3_col_wheel_front_right_02 on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
``` 

is this the way to test this ?

**Review by maltewi — APPROVED** [review]

**Comment by maltewi** [comment]
> manual reconfiguration test status: i tried removing the dynamics port
> 
>     1. i launch a CND with config 1 and got the ports generated
> 
>     2. when i launch another CND with config 2, i can get the port but also getting an error
> 
> 
> ```
> [20240209-13:53:26:220] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.state on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
> [20240209-13:53:34:440] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.state on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
> [20240209-13:53:34:440] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.coyote3_col_wheel_front_left_01 on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
> [20240209-13:53:34:440] [ERROR] - cnd/execution::Failed to create log port coyote3_chain_publisher.coyote3_col_wheel_front_right_02 on task coyote3_chain_publisher_Logger. (/opt/workspace/tools/cnd/execution/src/LogManager.cpp:106 - void cnd::execution::LogManager::taskAboutToStart(const string&))
> ```
> 
> is this the way to test this ?

This is unrelated to your patch. See https://git.hb.dfki.de/dfki-control/modeling/cnd-execution/-/issues/26
